### DISCUSSION TITLE
Are the Client Components comparable to e. g. Islands in Fresh?

### DISCUSSION THREAD

**Discussion by FleetAdmiralJakob** [vercel/next.js#52097] [App Router]
Now with server components as standard.

**Comment by icyJoseph** [comment]
In a way yes. If you have a React tree with nothing but Server Components, no JS code will be shipped for that tree. 

> Note that Next.js' implementation has some router, and other supporting, code that is still shipped, but they tried to keep it as low as possible.

As you start to include Client Components, which are the classic components we've always had, the JavaScript functions that tell React how these components describe the UI, are sent to the client.

React knows how to weave the server side rendered HTML with the interactivity on the client. 

> It always did, but it was all or nothing hydration, now it is selective 

There's basically, sections of the React tree which are static HTML strings, and others which have interactivity and can update their DOM representations at runtime.

Also, Client Components are still SSR'd - and more than a standard, Server Components are a default. So that, if you don't use interactivity, the framework, by default doesn't ship JS for that part. However if the component itself, or its importer are using the `'use client'` directive, then you opt-in into the old behaviour, of sending JS to the client, only for this section though.

- https://github.com/reactwg/server-components/discussions/4
- https://nextjs.org/docs/getting-started/react-essentials#client-components:
- https://nextjs.org/docs/getting-started/react-essentials:

Composition also plays a nice role here, because since with React, we can declare "slots", or "holes", using the `children` for example,  where other components will be placed, then a Client Component can make use of that to declare slots for Server Components, and then React knows also how to slot these together, https://nextjs.org/docs/getting-started/react-essentials#recommended-pattern-passing-server-components-to-client-components-as-props
